packages/headless/src/controllers/triggers/headless-query-trigger.ts

```diff
@@ -0,0 +1,82 @@
+import {Engine} from '../../app/headless-engine';
+import {ConfigurationSection, TriggerSection} from '../../state/state-sections';
+import {configuration, triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {executeSearch} from '../../features/search/search-actions';
+import {
+  updateQuery,
+  UpdateQueryActionCreatorPayload,
+} from '../../features/query/query-actions';
+import {logTriggerQuery} from '../../features/triggers/trigger-analytics-actions';
+
+/**
+ * The `QueryTrigger` controller handles query triggers.
+ */
+export interface QueryTrigger extends Controller {
+  /**
+   * the state of the `QueryTrigger` controller.
+   */
+  state: QueryTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `QueryTrigger` controller.
+ */
+export interface QueryTriggerState {
+  /**
+   * The new query to perform a search with after receiving a query trigger.
+   */
+  query: string | null;
+}
+
+/**
+ * Creates a `QueryTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `QueryTrigger` controller instance.
+ * */
+export function buildQueryTrigger(engine: Engine<object>): QueryTrigger {
+  if (!loadQueryTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        if (this.state.query) {
```

Comment:
It's a smaller detail, but I'd prefer to avoid using the `this` keyword because the state getter can be overridden. 

```
const trigger = buildQueryrTrigger(...)
const enhancedTrigger = {
     ...trigger,
    get state() {
      return // an object that does not have a "query" key
    }
}
```
It's unlikely that someone would do this, but why take the risk,